src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/CodeStyleOption2`1.cs

```diff
@@ -0,0 +1,214 @@
+            {
+                // data from storage is corrupt, or nothing has been stored yet.
+                return Default;
+            }
+
+            if (version != SerializationVersion)
+            {
+                return Default;
+            }
+
+            var parser = GetParser(typeAttribute.Value);
+            var value = parser(valueAttribute.Value);
+            var severity = (DiagnosticSeverity)Enum.Parse(typeof(DiagnosticSeverity), severityAttribute.Value);
+
+            return new CodeStyleOption2<T>(value, severity switch
+            {
+                DiagnosticSeverity.Hidden => NotificationOption2.Silent,
+                DiagnosticSeverity.Info => NotificationOption2.Suggestion,
+                DiagnosticSeverity.Warning => NotificationOption2.Warning,
+                DiagnosticSeverity.Error => NotificationOption2.Error,
+                _ => throw new ArgumentException(nameof(element)),
+            });
+        }
+
+        private static Func<string, T> GetParser(string type)
+            => type switch
+            {
+                nameof(Boolean) =>
+                    // Try to map a boolean value.  Either map it to true/false if we're a 
+                    // CodeStyleOption<bool> or map it to the 0 or 1 value for an enum if we're
+                    // a CodeStyleOption<SomeEnumType>.
+                    (Func<string, T>)(v => Convert(bool.Parse(v))),
+                nameof(Int32) => v => Convert(int.Parse(v)),
+                nameof(String) => v => (T)(object)v,
+                _ => throw new ArgumentException(nameof(type)),
+            };
+
+        private static T Convert(bool b)
+        {
+            // If we had a bool and we wanted a bool, then just return this value.
+            if (typeof(T) == typeof(bool))
+            {
+                return (T)(object)b;
+            }
+
+            // Map booleans to the 1/0 value of the enum.
+            return b ? (T)(object)1 : (T)(object)0;
+        }
+
+        private static T Convert(int i)
+        {
+            // We got an int, but we wanted a bool.  Map 0 to false, 1 to true, and anything else to default.
+            if (typeof(T) == typeof(bool))
+            {
+                return (T)(object)(i == 1);
+            }
+
+            // If had an int and we wanted an enum, then just return this value.
+            return (T)(object)(i);
+        }
+
+        public bool Equals(CodeStyleOption2<T> other)
+            => EqualityComparer<T>.Default.Equals(Value, other.Value) &&
+               Notification == other.Notification;
+
+        public override bool Equals(object obj)
+            => obj is CodeStyleOption2<T> option &&
+               Equals(option);
+
+        public override int GetHashCode()
+            => unchecked((Notification.GetHashCode() * (int)0xA5555529) + Value.GetHashCode());
+
+#if !CODE_STYLE
+        public static implicit operator CodeStyleOption<T>(CodeStyleOption2<T> option)
```

Comment:
The part within this `if !CODE_STYLE` is newly added code for implicit conversions between the internal and public types. This allows easy use of either of these types in Workspace and above layers, with ease of calling into shared code via conversions. We may still want to just stick with one set of types through Roslyn, but these conversions give us flexibility. Note the code within this `#if` can also be potentially split into a separate partial declaration and moved into Workspaces project.

Reply:
I'm not a fan of the implicit conversions. I would much rather see explicit operations (via extension methods) considering most of the code was already updated to consolidate on the new types as part of this pull request.

Reply:
Sounds good to me. I'll convert all implicit conversions added in this PR to extension methods.

Reply:
I switched these to explicit conversion operators - it was not clear if the extension method naming would be clear enough, additionally it is unnecessarily verbose. Explicit operators ensure that there is explicit intent with a cast expression. Let's discuss tomorrow if we prefer implicit operators, explicit operators or extension methods.